### PR TITLE
Add signup endpoint

### PR DESCRIPTION
**Pull request by ruihildt**
# Description

Add the signup endpoint.

## Checklist

- [x] I have performed a self-review of my own code

### CODE REVIEW

File: api/controllers/user.js [hunk 1]
```diff
@@ -36,5 +36,22 @@ module.exports = {
     } catch (error) {
       return response.error(res, 500, error.message);
     }
+  },
+
+  async createUser(req, res) {
+    let user = req.body;
+
+    let hash = bcrypt.hashSync(user.password, 10);
```
Comment (iamcristos) [inline api/controllers/user.js]: use const hash;
line 42

File: api/controllers/user.js [hunk 2]
```diff
@@ -36,5 +36,22 @@ module.exports = {
     } catch (error) {
       return response.error(res, 500, error.message);
     }
+  },
+
+  async createUser(req, res) {
+    let user = req.body;
```
Comment (iamcristos) [inline api/controllers/user.js]: why using let; I can't find where you are changing the value

File: api/controllers/user.js [hunk 3]
```diff
@@ -36,5 +36,22 @@ module.exports = {
...
+
+    let hash = bcrypt.hashSync(user.password, 10);
+    user.password = hash;
+
+    try {
+      const user = await models.Users.create(user);
+      if (user) {
+        return response.success(res, 201, `${user.username} account created successfully`);
+      }
+      return response.error(res, 404, 'Could not create Profile');
```
Comment (iamcristos) [inline api/controllers/user.js]: the error status code should be 400;
404 means not found

File: api/middleware/localStrategy.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        where: {
+          'email': email
+        }
+      })
+      .then(user => {
+        if (!user) {
+          return cb(null, false, { message: 'Incorrect email or password.' });
+        }
+
+        let hashedPassword = bcrypt.hashSync(password, user.salt);
```
Comment (iamcristos) [inline api/middleware/localStrategy.js]: also, note the let here use const since you are not reassigning the value

File: api/controllers/user.js [hunk 5]
```diff
@@ -58,6 +60,42 @@ module.exports = {
...
 
+  async createUser(req, res) {
+    try {
+      const user = await models.Users.create(req.body);
+      if (user) {
+        return response.success(res, 201, user);
+      }
+      return response.error(res, 400, 'Could not create Profile');
+    } catch (error) {
+      console.log(error.message);
```
Comment (iamcristos) [inline api/controllers/user.js]: clear console.log

File: api/controllers/user.js [hunk 6]
```diff
@@ -58,6 +60,42 @@ module.exports = {
...
+
+  async loginUser(req, res, next) {
+    const { email, password } = req.body;
+
+    try {
+      const user = await models.Users.findOne({
+        where: { email },
+        attributes: ['password', 'email', 'username', 'id']
+      });
+      console.log(user.dataValues.password);
```
Comment (iamcristos) [inline api/controllers/user.js]: here too

File: api/validator/userValidator.js [hunk 7]
```diff
@@ -61,5 +62,41 @@ module.exports = {
...
+      if (!user) {
+        return next();
+      }
+      return response.error(
+        res,
+        400,
+        'User already registered with username or email provided'
+      );
+    } catch (error) {
+      console.log(error.message);
```
Comment (iamcristos) [inline api/validator/userValidator.js]: here also

File: database/models/users.js [hunk 8]
```diff
@@ -46,10 +42,6 @@ module.exports = (sequelize, DataTypes) => {
           isEmail: {
             args: true,
             msg: 'Please enter a valid email'
```
Comment (damola-sd) [inline database/models/users.js]: The isUnique description has been removed. Why?

Comment (ruihildt) [reply database/models/users.js]: It has been removed because it's not a sequelize method. :dancer:

**Comment by iamcristos** [comment]
> Looks good. I have a comment about the isUnique tag in the database model. Does the isEmail make it unique already without needing to add it?

@damola-sd  those are model-based validations with validator apparently isUnique is not supported and we've taken that out, isEmail is there to ensure a valid email type is supplied

**Pull request merged**